app/src/main/java/com/babylon/wallet/android/presentation/settings/securitycenter/securityfactors/biometricspin/BiometricsPinViewModel.kt

```diff
@@ -0,0 +1,176 @@
+                    }
+                } else {
+                    _state.update { state ->
+                        state.copy(
+                            deviceFactorSources = state.deviceFactorSources.add(
+                                factorSourceCard
+                            )
+                        )
+                    }
+                }
+            }
+            .flowOn(defaultDispatcher)
+            .launchIn(viewModelScope)
+    }
+
+    @Suppress("UnusedParameter") // TODO
+    fun onDeviceFactorSourceClick(factorSourceId: FactorSourceId) {
+        viewModelScope.launch {
+            sendEvent(Event.NavigateToDeviceFactorSourceDetails)
+        }
+    }
+
+    private fun DeviceFactorSource.toFactorSourceCard(
+        includeDescription: Boolean = false,
+        messages: PersistentList<FactorSourceStatusMessage> = persistentListOf(),
+        accounts: PersistentList<Account> = persistentListOf(),
+        personas: PersistentList<Persona> = persistentListOf(),
+        hasHiddenEntities: Boolean
+    ): FactorSourceCard {
+        return FactorSourceCard(
+            id = id.asGeneral(),
+            name = hint.label,
+            includeDescription = includeDescription,
+            lastUsedOn = common.lastUsedOn.relativeTimeFormatted(),
+            kind = kind,
+            messages = messages,
+            accounts = accounts,
+            personas = personas,
+            hasHiddenEntities = hasHiddenEntities
+        )
+    }
+
+    private suspend fun getSecurityPromptsForDeviceFactorSource(
+        deviceFactorSourceId: FactorSourceId,
+        entitiesLinkedToDeviceFactorSource: EntitiesLinkedToFactorSource
+    ): PersistentList<FactorSourceStatusMessage> {
+        val isDeviceFactorSourceLinkedToAnyEntities = listOf(
+            entitiesLinkedToDeviceFactorSource.accounts,
+            entitiesLinkedToDeviceFactorSource.personas,
+            entitiesLinkedToDeviceFactorSource.hiddenAccounts,
+            entitiesLinkedToDeviceFactorSource.hiddenPersonas
+        ).any { it.isNotEmpty() }
+
+        val backedUpFactorSourceIds = preferencesManager.getBackedUpFactorSourceIds().firstOrNull().orEmpty()
+
+        return if (isDeviceFactorSourceLinkedToAnyEntities) {
+            val deviceFactorSourceIntegrity = entitiesLinkedToDeviceFactorSource.integrity as FactorSourceIntegrity.Device
+            deviceFactorSourceIntegrity.toMessages().toPersistentList()
+        } else if (backedUpFactorSourceIds.contains(deviceFactorSourceId)) { // if not linked entities we can't check
+            // the integrity, but we can check if the user backed up the seed phrase
+            persistentListOf(FactorSourceStatusMessage.NoSecurityIssues)
+        } else {
+            // otherwise we don't show any warnings
+            persistentListOf()
+        }
+    }
+
+    private fun FactorSourceIntegrity.Device.toMessages(): List<FactorSourceStatusMessage> {
```

Comment:
There must be some problem with this because I restored a wallet and I introduced the seed phrase during, but it shows the "Write down seed phrase.." message anyway.

Reply:
yes. Please check the description:

> ⚠️ The Biometrics/PIN screen incorrectly displays the "write down seed phrase" warning due to a [mapping](https://github.com/radixdlt/sargon/blob/6a4b46e1223910041d420e617be2fbe796b3a7c4/jvm/sargon-android/src/main/java/com/radixdlt/sargon/os/driver/AndroidStorageDriver.kt#L88) issue in Android Sargon OS. However, we decided with @GhenadieVP not to fix it, as the warning is now obsolete, and a refactor/overhaul of the security problems will follow.

Reply:
Oh, I've read the description but totally forgot to consider it when reviewing the code. Sorry!